tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
-            channels = discovery_info["channels"]
-            component = discovery_info["component"]
-            key = (component, unique_id)
-            entity_id = entity_registry.async_get_entity_id(component, "zha", unique_id)
-
-            assert key in entity_map
-            assert entity_id is not None
-            no_tail_id = NO_TAIL_ID.sub("", entity_map[key]["entity_id"])
-            assert entity_id.startswith(no_tail_id)
-            assert set([ch.name for ch in channels]) == set(entity_map[key]["channels"])
+
+def test_discover_by_device_type_override():
+    """Test entity discovery by device type overriding."""
+
+    ep_channels = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ep_mock = mock.PropertyMock()
+    ep_mock.return_value.profile_id = 0x0104
+    ep_mock.return_value.device_type = 0x0100
+    type(ep_channels).endpoint = ep_mock
+
+    overrides = {ep_channels.unique_id: {"type": zha_const.SWITCH}}
+    get_entity_mock = mock.MagicMock(
+        return_value=(mock.sentinel.entity_cls, mock.sentinel.claimed)
+    )
+    with mock.patch(
+        "homeassistant.components.zha.core.registries.ZHA_ENTITIES.get_entity",
+        get_entity_mock,
+    ):
+        with mock.patch.dict(disc.PROBE._device_configs, overrides, clear=True):
+            disc.PROBE.discover_by_device_type(ep_channels)
+            assert get_entity_mock.call_count == 1
+            assert ep_channels.claim_channels.call_count == 1
+            assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+            assert ep_channels.async_new_entity.call_count == 1
+            assert ep_channels.async_new_entity.call_args[0][0] == zha_const.SWITCH
+            assert (
+                ep_channels.async_new_entity.call_args[0][1] == mock.sentinel.entity_cls
+            )
+
+
+def test_discover_probe_single_cluster():
+    """Test entity discovery by single cluster."""
+
+    ep_channels = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ep_mock = mock.PropertyMock()
+    ep_mock.return_value.profile_id = 0x0104
+    ep_mock.return_value.device_type = 0x0100
+    type(ep_channels).endpoint = ep_mock
+
+    get_entity_mock = mock.MagicMock(
+        return_value=(mock.sentinel.entity_cls, mock.sentinel.claimed)
+    )
+    channel_mock = mock.MagicMock(spec_set=base_channels.ZigbeeChannel)
+    with mock.patch(
+        "homeassistant.components.zha.core.registries.ZHA_ENTITIES.get_entity",
+        get_entity_mock,
+    ):
+        disc.PROBE.probe_single_cluster(zha_const.SWITCH, channel_mock, ep_channels)
+
+    assert get_entity_mock.call_count == 1
+    assert ep_channels.claim_channels.call_count == 1
+    assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+    assert ep_channels.async_new_entity.call_count == 1
+    assert ep_channels.async_new_entity.call_args[0][0] == zha_const.SWITCH
+    assert ep_channels.async_new_entity.call_args[0][1] == mock.sentinel.entity_cls
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)